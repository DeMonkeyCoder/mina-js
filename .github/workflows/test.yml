name: Run tests
on:
    push:
        branches:
            - main
jobs:
    my-job:
        name: my-job
        runs-on: ubuntu-latest
        steps:
            - uses: actions/checkout@v4
            - uses: oven-sh/setup-bun@v2
            - run: bun install
            - run: bun run build
            - run: bun run test
