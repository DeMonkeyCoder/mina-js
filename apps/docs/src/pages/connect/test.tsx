import { TestZkApp } from "../../components/test-zkapp";

const TestPage = () => {
	if (import.meta.env.SSR) return null;
	return <TestZkApp />;
};

export default TestPage;
