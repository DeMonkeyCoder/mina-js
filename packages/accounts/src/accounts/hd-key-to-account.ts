import { sha256 } from "@noble/hashes/sha256";
import { bytesToHex } from "@noble/hashes/utils";
import { base58check } from "@scure/base";
import {
	type HDAccount,
	type HDKey,
	type HDOptions,
	MinaKeyConst,
} from "../types";
import {
	type PrivateKeyToAccountParams,
	privateKeyToAccount,
} from "./private-key-to-account";

export type HDKeyToAccountOptions = HDOptions &
	Partial<PrivateKeyToAccountParams>;

/**
 * @description Creates an Account from a HD Key.
 *
 * @returns A HD Account.
 */
export function hdKeyToAccount(
	hdKey_: HDKey,
	{
		accountIndex = 0,
		addressIndex = 0,
		changeIndex = 0,
		path,
		...options
	}: HDKeyToAccountOptions = {},
): HDAccount {
	const childNode = hdKey_.derive(
		path ||
			`m/${MinaKeyConst.PURPOSE}'/${MinaKeyConst.MINA_COIN_TYPE}'/${accountIndex}'/${changeIndex}/${addressIndex}`,
	);
	if (!childNode.privateKey) throw new Error("Private key not found");
	const childPrivateKeyArray = new Uint8Array(childNode.privateKey);
	childPrivateKeyArray[0] &= 0x3f;
	const childPrivateKey = childPrivateKeyArray.reverse();
	const privateKeyHex = `5a01${bytesToHex(childPrivateKey)}`;
	if (!privateKeyHex) throw new Error("Private key is empty");
	if (!privateKeyHex.match(/.{1,2}/g))
		throw new Error("Failed to split privateKeyHex into bytes");
	const privateKeyBytes = new Uint8Array(
		privateKeyHex.match(/.{1,2}/g)?.map((byte) => Number.parseInt(byte, 16)) ||
			[],
	);
	const account = privateKeyToAccount({
		...options,
		privateKey: base58check(sha256).encode(privateKeyBytes),
	});
	return {
		...account,
		getHdKey: () => childNode,
		source: "hd",
	};
}
