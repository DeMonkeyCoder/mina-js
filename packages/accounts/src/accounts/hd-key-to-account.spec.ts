import { describe, expect, it } from "bun:test";
import { hex } from "@scure/base";
import { HDKey } from "..";
import { hdKeyToAccount } from "./hd-key-to-account";

const hdKey = HDKey.fromMasterSeed(
	hex.decode(
		"59eabf9e91adfc40f7529e619fd9a7b47d41621715c27ce699da9a3a741536f377c405e496c91361ff3d65de1bb60f52c0add214d186ea0721151bee914e55fb",
	),
);

it("matches the snapshot", () => {
	const hdAccount = hdKeyToAccount(hdKey);
	expect(hdAccount).toMatchSnapshot();
});

describe("args: addressIndex", () => {
	Array.from({ length: 10 }).forEach((_, index) => {
		it(`derives addressIndex: ${index}`, () => {
			const account = hdKeyToAccount(hdKey, {
				addressIndex: index,
			});
			expect(account.publicKey.length).toEqual(55);
			expect(account.publicKey).toStartWith("B62");
		});
	});
});

describe("args: path", () => {
	Array.from({ length: 10 }).forEach((_, index) => {
		it(`derives path: m/44'/12586'/0'/0/${index}`, () => {
			const account = hdKeyToAccount(hdKey, {
				path: `m/44'/12586'/0'/0/${index}`,
			});
			expect(account.publicKey.length).toEqual(55);
			expect(account.publicKey).toStartWith("B62");
		});
	});
});

it("derives with accountIndex", () => {
	const hdAccount = hdKeyToAccount(hdKey, { accountIndex: 1 }).publicKey;
	expect(hdAccount).toMatchSnapshot();
});

it("derives with changeIndex", () => {
	const hdAccount = hdKeyToAccount(hdKey, { changeIndex: 1 }).publicKey;
	expect(hdAccount).toMatchSnapshot();
});

it("signs a message", async () => {
	const account = hdKeyToAccount(hdKey);
	const signature = await account.signMessage({ message: "hello word" });
	expect(signature).toMatchSnapshot();
});
