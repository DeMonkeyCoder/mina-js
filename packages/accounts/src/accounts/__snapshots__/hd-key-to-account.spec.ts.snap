// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
{
  "createNullifier": [Function: AsyncFunction],
  "getHdKey": [Function],
  "publicKey": "B62qkYa1o6Mj6uTTjDQCob7FYZspuhkm4RRQhgJg9j4koEBWiSrTQrS",
  "signFields": [Function: AsyncFunction],
  "signMessage": [Function: AsyncFunction],
  "signTransaction": [Function: AsyncFunction],
  "source": "hd",
  "type": "local",
}
`;

exports[`derives with accountIndex 1`] = `"B62qpmbhZAQHHQHSMwngKy7NuNSEQav7GeTPKRQ9FSdPChAnYiwWyLz"`;

exports[`derives with changeIndex 1`] = `"B62qoxndFokiZ2K4uCNV8673xV3V2qPcmZuQKHBQ1Zm4Narz7keS5bj"`;

exports[`signs a message 1`] = `
{
  "data": "hello word",
  "publicKey": "B62qkYa1o6Mj6uTTjDQCob7FYZspuhkm4RRQhgJg9j4koEBWiSrTQrS",
  "signature": {
    "field": "1737565114197582632206802090873583593478816160361909498260553982818075524406",
    "scalar": "8097562446343817256257278438860076058640076299120594151149598887065384496675",
  },
}
`;
